### PR TITLE
Replace hyperscan by vectorscan

### PR DESCRIPTION
**Pull request by rschu1ze**
This commit migrates ClickHouse to Vectorscan. The first 10 min of
[0] explain the reasons for it.

(*) Addresses (but does not resolve) https://github.com/ClickHouse/ClickHouse/issues/38046

(*) Names of hyperscan-related config parameters (e.g.
    "max_hyperscan_regexp_length") are preserved for compatibility.
    Likewise, error codes (e.g.
    "ErrorCodes::HYPERSCAN_CANNOT_SCAN_TEXT") and function/class names
    (e.g. "HyperscanDeleter") are preserved as vectorscan aims to be a
    drop-in replacement.

[0] https://www.youtube.com/watch?v=KlZWmmflW6M

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Migrate from the Intel hyperscan library to vectorscan, this speeds up many string matching on non-x86 platforms

### CODE REVIEW

**Comment by qoega** [comment]
Why we need change name of a setting? I think is just will add confusion.

**Comment by rschu1ze** [comment]
We don't need to. Vectorscan tries to stay API/ABI-compatible with hyperscan. Will revert the cfg parameter renaming.

**Comment by alexey-milovidov** [comment]
> The first 10 min of [0] explain the reasons for it.

I don't see the link.

File: contrib/hyperscan-cmake/CMakeLists.txt [hunk 1]
```diff
@@ -1,239 +1,232 @@
...
-    PRIVATE -g0 # Library has too much debug information
-    -mno-avx -mno-avx2 # The library is using dynamic dispatch and is confused if AVX is enabled globally
-    -march=corei7 -O2 -fno-strict-aliasing -fno-omit-frame-pointer -fvisibility=hidden # The options from original build system
-    -fno-sanitize=undefined # Assume the library takes care of itself
-)
-target_include_directories (_hyperscan
-    PRIVATE
-        common
-        "${LIBRARY_DIR}/include"
+    "${LIBRARY_DIR}/src/util/supervector/arch/arm/impl.cpp"
```
Comment (alexey-milovidov) [inline contrib/hyperscan-cmake/CMakeLists.txt:210]: Most likely we cannot unconditionally compile all three?

Comment (rschu1ze) [reply contrib/hyperscan-cmake/CMakeLists.txt:210]: True. This was just the very first try. The goal here is to compile x86, arm and ppc separately.

By the way, vectorscan and hyperscan diverged quite a bit already. The former also has a new dependency on "Ragel" (http://www.colm.net/open-source/ragel/). I'll check but perhaps we need a fourth-party lib to make the third-party lib work.

Comment (alexey-milovidov) [reply contrib/hyperscan-cmake/CMakeLists.txt:210]: We can add ragel to contrib at the top level and make vectorscan use it.

Comment (danlark1) [reply contrib/hyperscan-cmake/CMakeLists.txt]: There are only 2 files of ragel, back then I just invoked ragel on them not to add extra dependency

**Comment by rschu1ze** [comment]
About ragel: I went with @danlark1 proposal and saved the external dep.

About cross-porting fixes from hyperscan to vectorscan: https://github.com/VectorCamp/vectorscan/issues/112 😄 Thanks! I'm happy to update vectorscan in ClickHouse once the fixes are in.

**Comment by rschu1ze** [comment]
Ah ... some tests no longer ran after the switch to VS. Re-enabled them.

About CI test failures:
- I ran all statless tests for functions `multiMatch`, `multiMatchAnyIndex`, `multiFuzzyMatchIndex`, `multiFuzzyMatchAnyIndex` locally on ARM M1 and they were green
- the timeout of the fuzzer test in CI looks unrelated
- similarly, the many integration test failures in CI look unrelated
- `02226_or_like_combine.sql` fails but it's not really clear to me if the error is related to vectorscan. Locally, the test was green.
- the style check complained about typos but the file listing the typos was empty.

Guess it makes sense to merge from master, perhaps some of the errors were fixed upstream.

**Comment by rschu1ze** [comment]
I can reproduce a weird error on AWS EC2 Graviton that vectorscan executed from multiMatchAny() 1. in test "00926_multimatch" fails to parse cyrillic characters (= throws in release builds / crashes in dbg builds), 2. in test "02226_or_like_combine" produces wrong results with cyrillic characters, whereas hyperscan had no problem with either test. This happens only on ARM (for which vectorscan is enabled in this PR). I don't recall any i18n-related option, neither in the original build system nor in the bundled build script, in fact the bundled build script for x86 and ARM are pretty similar apart from compiler flags of course, and the implementation of multiMatchAny() is platform-agnostic.

So to me, this looks like a bug in vectorscan. Still think it makes sense to move ClickHouse to vectorscan but only on x86 which inherits the mature x86 code from hyperscan. Will disable ARM for now and check back in a separate PR.

**Comment by rschu1ze** [comment]
Oh, and perhaps someone likes to approve 😄. Thanks.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by danlark1** [comment]
> I can reproduce a weird error on AWS EC2 Graviton that vectorscan executed from multiMatchAny() 1. in test "00926_multimatch" fails to parse cyrillic characters (= throws in release builds / crashes in dbg builds), 2. in test "02226_or_like_combine" produces wrong results with cyrillic characters, whereas hyperscan had no problem with either test. This happens only on ARM (for which vectorscan is enabled in this PR). I don't recall any i18n-related option, neither in the original build system nor in the bundled build script, in fact the bundled build script for x86 and ARM are pretty similar apart from compiler flags of course, and the implementation of multiMatchAny() is platform-agnostic.
> 
> So to me, this looks like a bug in vectorscan. Still think it makes sense to move ClickHouse to vectorscan but only on x86 which inherits the mature x86 code from hyperscan. Will disable ARM for now and check back in a separate PR.

What's the locale on Gravitons? I managed to reproduce the issue when the machine does not have any locales set up, only POSIX one.

**Comment by alexey-milovidov** [comment]
Our builds should be hermetic - not depend on any locales in the system.

**Comment by rschu1ze** [comment]
@danlark1 Interesting, thanks!

With my custom dotfiles deployed on the Gravitons, the issue occurs with LC_ALL='en_US.UTF-8', not only with LC_ALL=C, LC_ALL=POSIX, or LC_ALL= (not set) - I verified that. Furthermore, the issue occurred in the CI's "stateless" tests on ARM but not on x86 and I assume the locale was configured equally on both platforms.

But maybe I did not exactly follow your steps to reproduce the issue. Would you like to share them? I.e. are you on x86 and did you just set LC_ALL or something else too?

If this is really a problem caused by locales, I agree with Alexey that the next step would be to find a way to force Unicode for vectorscan.

To be more specific about the repro:
- in test "00926_multimatch", query `SELECT 1 = multiMatchAny('фабрикант', ['f[ae]b[ei]rl', 'ф[иаэе]б[еэи][рпл]', 'афиукд', 'a[ft],th', '^ф[аиеэ]?б?[еэи]?$', 'берлик', 'fab', 'фа[беьв]+е?[рлко]']);` throws/crashes in vectorscans parser.
-  in test "02226_or_like_combine", query `SELECT materialize('Привет, optimized World') AS s WHERE (s LIKE 'hell%') OR (s ILIKE '%привет%') OR (s LIKE 'world%') SETTINGS optimize_or_like_chain = 1 SETTINGS allow_hyperscan = X;` produces with X set to 0 one result row (this is the expected result) and with X set to 1 no result rows.

**Comment by stevecapperarm** [comment]
Hello,
I managed to get both `00926_multimatch` and `02226_or_like_combine` to work by regenerating both of the "rageled" assets (`Parser.cpp` and `control_verbs.cpp`). There have been bugs in both colm and ragel on Arm platforms in the past (one notable one around the handling of chars, which are un-signed on Arm but signed on x86). For reference, I used upstream colm (`d00ef7930a9d42469c19a5880744a086065ee20c`) and upstream ragel (`d4577c924451b331c73c8ed0af04f6efd35ac0b4`).

I *haven't* tested this on x86 though.

Cheers,
Steve

**Comment by rschu1ze** [comment]
@stevecapperarm Very interesting, thanks!

I tried to do as you said in #41033 ... let's see what the tests say.

**Comment by stevecapperarm** [comment]
Thanks @rschu1ze ! Hmmm, I can see some build failures for AArch64 (due to sign problems). I suspect the rageled output may actually depend on the host used to generate it. (FWIW, I used an AArch64 host).

**Comment by rschu1ze** [comment]
Just checked ... ragel produces on x86 and arm hosts the same outputs (also just tried on my AArch64 host)

The [compile errors](https://s3.amazonaws.com/clickhouse-builds/41033/4d07e3ef8fa0a0010375aa4ad0d6b2a5ecb07677/binary_aarch64/build_log.log) are caused by ARM's vs. x86's different interpretation what a `char` is. Since this is 3rd party code, I am tempted to disable the compiler warning.

(I also made the mistake that in the first push I generated the assets from the vectorscan HEAD and not from ClickHouse's vectorscan fork - which lags a bit behind HEAD. But that didn't matter, both produced the same .cpp files.)

**Comment by rschu1ze** [comment]
[Test results on ARM](https://s3.amazonaws.com/clickhouse-test-reports/41033/c4297c6c997056531198dc0b4b6ebba10ed9d65f/stateless_tests__aarch64_.html) are promising, 02226_or_like_combine and 00926_multimatch are green.

**Pull request merged**